### ISSUE TITLE
MLPerf evaluation details

### ISSUE THREAD

**Issue by luiza-amd** [AMD-AGI/Primus#703]
### 1. Quality Target (Convergence Criterion)

For the **Flux.1** text-to-image benchmark (v6.0, replacing Stable Diffusion v2 from v5.0):

- **Quality metric**: **Eval loss**
- **Quality target**: **0.586 eval loss** (the model must reach this threshold)
- **Evaluation frequency**: Every **262,144 training samples** (`CEIL(262144 / global_batch_size) * global_batch_size` if not evenly divisible)

This is different from the old Stable Diffusion v2 benchmark which used FID <= 90 and CLIP >= 0.15. Flux uses a simpler **eval loss** convergence target.

### 2. Time To Train (TTT) -- Individual Run

A single **run result** is defined as the **wall-clock time** from start to convergence:

- Clock starts before any part of the system touches the dataset (up to 30 min of pre-clock model initialization is allowed in Closed division)
- Clock stops when any part of the system determines the quality target (0.586 eval loss) has been reached
- Clock **cannot be paused** during the run

### 3. From Multiple Runs to a Benchmark Result

For Flux, **10 runs** are required per submission. The scoring uses **"olympic scoring"**:

1. Run the benchmark **10 independent times**
2. **Drop the fastest and slowest** results (a single non-converging run can count as the slowest)
3. **Take the mean** of the remaining 8 run times
4. More than 1 non-converging run invalidates the entire submission

### 4. Normalization

The raw benchmark result (mean TTT) is then **normalized** against the MLPerf reference result:

```
Normalized Score = Reference Result / Benchmark Result
```

Higher numbers = better (faster than reference). This makes the score intuitive -- a score of 2.0 means 2x faster than reference.

### 5. RCP (Reference Convergence Points) Check and Adjustment

Before accepting the score, convergence is validated against **Reference Convergence Points**:

- The reference implementation provides convergence data at various batch sizes
- A **one-sided Student's t-test** (p=0.05) checks that the submission's mean epochs-to-converge isn't suspiciously faster than the reference
- If the submission converges faster but within the acceptable range, the score gets **normalized to the RCP mean**:

```
Normalized Score = Olympic Score * (RCP Mean Epochs / Submission Olympic Epochs)
```

This prevents submissions from benefiting from lucky convergence.

### 6. Summary Result (Optional)

MLCommons endorses a methodology for **custom summary results** across multiple benchmarks:

```
Custom Summary = Weighted Geometric Mean of normalized scores across benchmarks
```

This is optional -- MLPerf doesn't publish a single official summary number, but vendors often compute one across all benchmarks they submitted.

### Key Flux-Specific Rules

| Parameter | Value |
|-----------|-------|
| Model | FLUX.1 (11.9B params) |
| Dataset | CC12M (~12M samples) |
| Quality target | 0.586 eval loss |
| Eval frequency | Every 262,144 samples |
| Min runs | 10 |
| Epoch reporting | Training sample (integer) |
| Optimizer | AdamW (beta1=0.9, beta2=0.95, eps=1e-8, wd=0.1) |
| Gradient clip norm | 1.0 |
| Global batch size | Unconstrained |

### The Core Idea

Flux is a **flow-matching diffusion model**. During training, the model learns to predict the "velocity" (the direction from noise to clean image in latent space). Convergence is measured by how well the model predicts this velocity on a **held-out validation set** (COCO-2014), using a specific **eval loss** methodology inspired by the **Stable Diffusion 3 paper**.

### Step 1: What Happens During Training

Each training step:

1. Take a batch of images from **CC12M** dataset
2. Encode images into latent space using the **VAE autoencoder**
3. Encode text captions using **T5** and **CLIP** text encoders
4. Sample a random timestep `t ~ Uniform(0, 1)` for each sample
5. Create noisy latents: `latent_noisy = (1 - t) * latent_clean + t * noise`
6. The Flux transformer predicts the velocity: `v_pred = model(latent_noisy, t, text_embeddings)`
7. The loss is **MSE** between `v_pred` and the target `(noise - latent_clean)`

### Step 2: How Evaluation Works (the key part)

Every **262,144 training samples**, validation is triggered on a subset of **COCO-2014**. The eval loss is computed using **8 fixed equidistant timesteps** rather than random timesteps:

```
timesteps = [1/8 * (i + 0.5) for i in range(8)]
         = [0.0625, 0.1875, 0.3125, 0.4375, 0.5625, 0.6875, 0.8125, 0.9375]
```

This is the critical difference from training -- using fixed, deterministic timesteps gives a **stable and reproducible** quality signal. The SD3 paper showed this correlates highly with external metrics like FID and CLIP score.

Two strategies are allowed:

| Strategy | How it works | Trade-off |
|----------|-------------|-----------|
| **Round-robin** (default) | Each validation sample gets exactly 1 of the 8 timesteps in round-robin order. Over many samples, all timesteps are equally represented. | Faster, uses less memory |
| **All-timesteps** | Each validation sample is evaluated at all 8 timesteps (batch expanded 8x). | More precise per-sample, but 8x memory/compute |

Both converge to the same aggregate loss value over the full validation set.

### Step 3: Computing the Eval Loss Value

For each validation batch:

1. Load image + caption from COCO-2014 validation set
2. Encode image through VAE -> `latent_clean`
3. Encode text through T5 and CLIP -> `t5_encodings`, `clip_encodings`
4. Apply the assigned timestep(s):
   - `noise = randn_like(latent_clean)`
   - `sigma = timestep`
   - `latent_noisy = (1 - sigma) * latent_clean + sigma * noise`
5. Forward pass through the model -> `v_pred`
6. Compute MSE loss: `loss = MSE(v_pred, noise - latent_clean)`
7. Accumulate across all validation batches

After processing the full validation set:
```
eval_loss = sum(batch_losses) / num_steps
```
This is then averaged across all data-parallel ranks to get a **global average eval loss**.

### Step 4: Convergence Check

After each evaluation:

- If `eval_loss <= 0.586` --> **converged**, stop the clock
- If `eval_loss > 0.586` --> continue training for another 262,144 samples, then evaluate again

The wall-clock time from start to this moment is the **TTT (Time To Train)** for that run.

### Step 5: From TTT to Final Rating

Once you have the TTT from a single run, the multi-run scoring procedure applies:

```
10 independent runs
    --> drop fastest and slowest
    --> mean of remaining 8 = "olympic score" (raw TTT)
    --> check against RCP (reference convergence points)
    --> if converged faster than reference (but within acceptable range):
        normalized_score = olympic_score * (RCP_mean_epochs / submission_epochs)
    --> final_rating = reference_TTT / normalized_TTT
```
